### PR TITLE
Phpspec body params fix

### PR DESCRIPTION
**Pull request by echosa**
The body params in requests are causing phpspec to fail. I remember something similar before. If we remove the body params from the calls in both the failing specs and the code, the spec passes, so body params are definitely the issue.

What's happening is that the phpspec is expecting the Request objects to be _exactly_ the same object, not just two objects with the same values. However, this is not what is currently happening, since the Request in the spec and the Request in the code are both generated in with individual `new` calls, resulting in different objects in memory:

```
object(GuzzleHttp\Psr7\Request)#368 (7) {
```

vs

```
object(GuzzleHttp\Psr7\Request)#618 (7) {
```

Notice the IDs are different: 368 vs 618. These do not match, so the objects are different, which makes phpspec go "this call isn't the exact call I was told to expect".

I was able to prove this by setting the Request instantiated in a failing spec to `$_SESSION` (_ew_) and then changing the code to pull that Request and use it. This caused the spec to pass.

Basically, we're telling phpspec to expect a call like:

```
$guzzleClient->send(Request#368)->willReturn($response);
```

but what's happening in the code is:

```
$guzzleClient->send(Request#618);
```

We didn't define what should happen for a `$guzzleClient->send(Request#618);` call, so we get null back, as stated in the error messages.

In this case, it seems like phpspec is doing its job. It has identified a class that is doing too much: `AbstractResource`. That class is handling calls the Twitch API and *also* handling Request object creation. Hence the need for `new` calls in both spec and code. That is why I'm making this PR. I'm pulling the Request object creating into a new class which can then be stubbed/mocked/whatever so that our specs don't have to call `new Request()` anymore (which is ultimately the issue). I've found that using `new` in a spec usually means you've done something wrong.

With this PR, I'm introducing `RequestGenerator` which pulls the previously mentioned Request object creation code out of `AbstractResource` into its own class. 

This will require a LOT of changes throughout the code and specs. However, it will also fix an issue we've had multiple times and make the code more SOLID.

### CODE REVIEW

**Comment by echosa** [comment]
So far, all I've done is add the base code changes and fix one particular spec class as an example. The rest of the specs will need to be updated as well.

**Comment by echosa** [comment]
For the record, this work is being broken off from #100

**Comment by brandinarsenault** [comment]
@echosa I don't mind helping on this, fantastic find by the way! Are you planning on continuing work right now, or am I fine to crack on from wherever?

**Comment by echosa** [comment]
@brandinarsenault I'm stopping now, after fixing the `AnalysticsApiSpec` and `BitsApiSpec`. Please, feel free to jump in and help. I have quite limited time for this. 

Updating the specs is pretty easy and straightforward. Shouldn't give you too much trouble. Note that you need to specify the empty query and body parameter array in the `$requestGenerator->generate()` calls in the specs.

Once all the specs are updated, the library code itself will also need to be updated to handle the changes made to `AbstractResource`.

**Comment by brandinarsenault** [comment]
Removing `it_should_get_users_and_include_email` and `it_should_get_users_by_everything_including_email` from `UsersApiSpec` as they're not needed based on reviewing the documentation.

**Review by brandinarsenault — APPROVED** [review]

**Comment by brandinarsenault** [comment]
@echosa I'm unable to request a review from you since this is your PR, but since I did most of the work, can you review the files and provide any comments before I merge this?

**Comment by echosa** [comment]
@brandinarsenault Have you used this library with the changed code? Having specs and tests pass is all good and well, but I'd like to know that someone has actually used the refactored code for real and that it works. :-)

**Comment by brandinarsenault** [comment]
@echosa will begin running some spot tests now. Just confirming, if I validate it works, I'm good to merge?

I don't plan on doing a release until I get this, the EventSub and a couple issues merged through PRs, and we'll do a large 5.0 breaking release.

**Comment by ThibaultVlacich** [comment]
I have tested against this branch on my project, and my limited use of endpoints worked perfectly fine.

I am using the GamesApi, the StreamsApi and the OauthApi.

**Comment by brandinarsenault** [comment]
I found an issue which I'm resolving now...continuing testing.

**Comment by brandinarsenault** [comment]
Tests are passing and library working as expected :) @echosa any comments before I merge?

File: src/NewTwitchApi/NewTwitchApi.php [hunk 1]
```diff
@@ -50,29 +50,31 @@ class NewTwitchApi
     private $videosApi;
     private $webhooksApi;
     private $webhooksSubscriptionApi;
+    private $requestGenerator;
 
     public function __construct(Client $helixGuzzleClient, string $clientId, string $clientSecret, Client $authGuzzleClient = null)
     {
+        $this->requestGenerator = new RequestGenerator();
```
Comment (echosa) [inline src/NewTwitchApi/NewTwitchApi.php:57]: Does this need to be a class field, or could it just be a local variable in the function?

**Review by echosa — COMMENTED** [review]
Looks good! Let one minor comment, but then feel free to merge. :-)

File: src/NewTwitchApi/NewTwitchApi.php (hunk 1 above)
Comment (brandinarsenault) [reply src/NewTwitchApi/NewTwitchApi.php:57]: Just to confirm, you'd prefer it to be `$requestGenerator = new RequestGenerator();`? If so, yes, this works. Please confirm and I'll merge.

**Comment by brandinarsenault** [comment]
@echosa I believe I've addressed your comment. Please confirm and I'll merge.

**Comment by echosa** [comment]
All good to go! :+1:

**Pull request merged**